fw/cache.c

```diff
@@ -329,18 +330,55 @@ tfw_cache_key_node(unsigned long key)
 }
 
 /**
- * Just choose any CPU for each node to use queue_work_on() for
- * nodes scheduling. Reserve 0th CPU for other tasks.
+ * Release node-cpu map.
  */
 static void
+tfw_release_node_cpus(void)
+{
+	int node;
+
+	for(node = 0; node < nr_online_nodes; node++) {
+		if(c_nodes[node].cpu)
+			kfree(c_nodes[node].cpu);
+	}
+	kfree(c_nodes);
+}
+
+/**
+ * Create node-cpu map to use queue_work_on() for nodes scheduling.
+ * 0th CPU is reserved for other tasks.
+ * At the moment we doesn't support CPU hotplug, so enumerate only online CPUs.
+ */
+static int
 tfw_init_node_cpus(void)
 {
-	int cpu, node;
+	int nr_cpus, cpu, node;
+
+	T_DBG2("nr_online_nodes: %d", nr_online_nodes);
+
+	c_nodes = kzalloc(nr_online_nodes * sizeof(CaNode), GFP_KERNEL);
+		if (!c_nodes) {
+			T_ERR( "Failed to allocate nodes map for cache work scheduler");
+			return -ENOMEM;
+	}
+	for_each_node_with_cpus(node) {
+		nr_cpus = nr_cpus_node(node);
+		T_DBG2("node: %d  nr_cpus: %d",node, nr_cpus);
+		c_nodes[node].cpu = kmalloc(nr_cpus * sizeof(int), GFP_KERNEL);
+		if (!c_nodes[node].cpu) {
+			T_ERR( "Failed to allocate a CPU %i for cache work scheduler", cpu);
+			tfw_release_node_cpus();
```

Comment:
This comment is not resolved! Add debug print to tfw_release_node_cpus! Then simulate situation when tdb_open fails, you can see that tfw_release_node_cpus is not called! There is a memory leak. You should add tfw_release_node_cpus after node_cpus_alloc_err label not here!